{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Bad key \"text.kerning_factor\" on line 4 in\n",
      "/Users/katekistler/anaconda3/envs/nextstrain/lib/python3.6/site-packages/matplotlib/mpl-data/stylelib/_classic_test_patch.mplstyle.\n",
      "You probably need to get an updated matplotlibrc file from\n",
      "https://github.com/matplotlib/matplotlib/blob/v3.1.3/matplotlibrc.template\n",
      "or from the matplotlib source distribution\n"
     ]
    }
   ],
   "source": [
    "import math\n",
    "import json\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio import AlignIO\n",
    "from Bio.Align import MultipleSeqAlignment\n",
    "from Bio.Align import AlignInfo\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from scipy import stats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#divergence is calculated from consensus sequence at first time point"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def separate_clades(cov, gene):\n",
    "    if cov == 'hku1':\n",
    "        clade_file = '../'+str(cov)+'/results/clades_full.json'\n",
    "    else:\n",
    "        if gene =='spike' or gene == 's1' or gene == 's2':\n",
    "            clade_file = '../'+str(cov)+'/results/clades_spike.json'\n",
    "        else:\n",
    "            clade_file = '../'+str(cov)+'/results/clades_full.json'\n",
    "    \n",
    "    clade_lists = []\n",
    "    with open(clade_file, \"r\") as clade_handle:\n",
    "        clades = json.load(clade_handle)\n",
    "        for node, v in clades['nodes'].items():\n",
    "            if 'NODE' not in node:\n",
    "                clade_lists.append({'clade':v['clade_membership'],\n",
    "                                   'strain':node})\n",
    "\n",
    "    clade_df = pd.DataFrame(clade_lists)\n",
    "    return clade_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def syn_nonsyn_histogram(cov, gene, clade):\n",
    "    #Find frequency of nonsyn and syn SNPs\n",
    "    input_file_root = '../'+str(cov)+'/auspice/seasonal_corona_'+str(cov)+'_'+str(gene)+'_root-sequence.json'\n",
    "    input_file_alignment = '../'+str(cov)+'/results/aligned_'+str(cov)+'_'+str(gene)+'.fasta'\n",
    "    metafile = '../'+str(cov)+'/results/metadata_'+str(cov)+'_'+str(gene)+'.tsv'\n",
    "    \n",
    "\n",
    "    meta = pd.read_csv(metafile, sep = '\\t')\n",
    "    meta.drop(meta[meta['date']=='?'].index, inplace=True)\n",
    "    meta.dropna(subset=['date'], inplace=True)\n",
    "    meta['year'] = meta['date'].str[:4].astype('int')\n",
    "    \n",
    "    \n",
    "    if clade!= None:\n",
    "        clade_df = separate_clades(cov, gene)\n",
    "        meta = meta.merge(clade_df, on='strain')\n",
    "        meta.drop(meta[meta['clade']!=clade].index, inplace=True)\n",
    "    \n",
    "    strains = meta['strain'].tolist()\n",
    "                \n",
    "    fraction_subs = []\n",
    "    fraction_syn = []\n",
    "    fraction_nonsyn = []\n",
    "\n",
    "    #calculate divergence at all subsequent time points\n",
    "    with open(input_file_alignment, \"r\") as aligned_handle:\n",
    "        for virus in SeqIO.parse(aligned_handle, \"fasta\"):         \n",
    "            if virus.id in strains:    \n",
    "\n",
    "                #check\n",
    "                if len(virus.seq) != len(root_seq):\n",
    "                    print(virus)\n",
    "                elif len(virus.seq) == len(root_seq):\n",
    "                    count_total_unambiguous = 0\n",
    "                    count_subs = 0\n",
    "                    count_syn_subs = 0\n",
    "                    count_nonsyn_subs = 0\n",
    "                    for pos in range(len(root_seq)):\n",
    "                        root_nt = str(root_seq[pos])\n",
    "                        virus_nt = str(virus.seq[pos])\n",
    "                        #skip ambiguous sites\n",
    "                        if virus_nt != 'N':\n",
    "                            if root_nt != 'N':\n",
    "                                count_total_unambiguous+=1\n",
    "                                if virus_nt != root_nt:\n",
    "                                    count_subs+=1\n",
    "                                    #determine syn or nonsyn\n",
    "                                    codon = math.floor(pos/3)\n",
    "                                    codon_pos = pos-(codon*3)\n",
    "                                    if codon_pos == 0:\n",
    "                                        codon_nt = virus.seq[pos:(pos+3)]\n",
    "                                    elif codon_pos == 1:\n",
    "                                        codon_nt = virus.seq[(pos-1):(pos+2)]\n",
    "                                    elif codon_pos == 2:\n",
    "                                        codon_nt = virus.seq[(pos-2):(pos+1)]\n",
    "                                    codon_aa = codon_nt.translate()\n",
    "                                    root_aa = root_aa_seq[codon]\n",
    "                                    #skip ambiguous\n",
    "                                    if root_aa != 'X':\n",
    "                                        if codon_aa != root_aa:\n",
    "                                            count_nonsyn_subs+=1\n",
    "                                            nt_mut = str(root_nt)+str(pos)+str(virus_nt)+'nonsyn'\n",
    "                                        elif codon_aa == root_aa:\n",
    "                                            count_syn_subs+=1\n",
    "                                            nt_mut = str(root_nt)+str(pos)+str(virus_nt)+'syn'\n",
    "                                        if nt_mut not in nt_muts:\n",
    "                                            nt_muts.append(nt_mut)\n",
    "                    fraction_sub_virus = count_subs/count_total_unambiguous\n",
    "                    fraction_subs.append(fraction_sub_virus)\n",
    "                    fraction_syn_virus = count_syn_subs/count_total_unambiguous\n",
    "                    fraction_syn.append(fraction_syn_virus)\n",
    "                    fraction_nonsyn_virus = count_nonsyn_subs/count_total_unambiguous\n",
    "                    fraction_nonsyn.append(fraction_nonsyn_virus)\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "oc43/KF572819/039A_07/2007\n",
      "oc43/KF572861/5617_07/2007\n",
      "oc43/KF572842/4795_07/2007\n",
      "oc43/KF572844/5240_07/2007\n",
      "oc43/KF572847/5352_07/2007\n",
      "oc43/KF572851/5445_07/2007\n",
      "oc43/KF572860/5595_07/2007\n",
      "oc43/KF572863/5656_07/2007\n",
      "oc43/KF572864/69A_07/2007\n",
      "oc43/KF572846/5345_07/2007\n",
      "oc43/KF572850/5442_07/2007\n",
      "oc43/KF572852/5472_07/2007\n",
      "oc43/KF572855/5485_07/2007\n",
      "oc43/KF572857/5517_07/2007\n",
      "oc43/KF572854/5484_07/2007\n",
      "oc43/KF572848/5370_07/2007\n",
      "oc43/KF572853/5479_07/2007\n",
      "oc43/LC506765/HCoV_OC43/Yamagata.JPN/11.2009_0765/2009\n",
      "oc43/KF572843/4954_07/2007\n",
      "oc43/KF572856/5508_07/2007\n",
      "oc43/KF572868/892A_08/2008\n",
      "oc43/KF572862/5625_07/2007\n",
      "oc43/KF572845/5331_07/2007\n",
      "oc43/KF572859/5566_07/2007\n",
      "oc43/KF572820/079A_07/2007\n",
      "oc43/KF572858/5519_07/2007\n",
      "oc43/KF572821/10108_10/2010\n",
      "oc43/KF572824/1034A_08/2008\n",
      "oc43/KF572826/1081A_08/2008\n",
      "oc43/KF572828/1157A_08/2008\n",
      "oc43/KF572830/1216A_08/2008\n",
      "oc43/KF572834/1357A_09/2009\n",
      "oc43/KF572837/1593A_09/2009\n",
      "oc43/KF572871/9138_09/2009\n",
      "oc43/KF572872/978A_08/2008\n",
      "oc43/KF572836/1591A_09/2009\n",
      "oc43/KF572835/1382A_09/2009\n",
      "oc43/KF572866/8164_09/2009\n",
      "oc43/KF572865/8099_09/2009\n",
      "oc43/KF572867/8375_09/2009\n",
      "oc43/KF572827/1135A_08/2008\n",
      "oc43/LC506768/HCoV_OC43/Yamagata.JPN/04.2010_0251/2010\n",
      "oc43/LC506770/HCoV_OC43/Yamagata.JPN/20.2010_1341/2010\n",
      "oc43/LC506771/HCoV_OC43/Yamagata.JPN/21.2010_1394/2010\n",
      "oc43/LC506769/HCoV_OC43/Yamagata.JPN/20.2010_1337/2010\n",
      "oc43/KF572839/2151A_10/2010\n",
      "oc43/KF572870/9001_09/2009\n",
      "oc43/KF572825/10574_10/2010\n",
      "oc43/KF572838/2134A_10/2010\n",
      "oc43/KF572822/10285_10/2010\n",
      "oc43/LC506767/HCoV_OC43/Yamagata.JPN/03.2010_0241/2010\n",
      "oc43/KF963242/HCoV_OC43/FRA_EPI/Caen/2011/13/2011\n",
      "oc43/LC506809/HCoV_OC43/Yamagata.JPN/14.2012_0781/2012\n",
      "oc43/LC506794/HCoV_OC43/Yamagata.JPN/14.2011_0897/2011\n",
      "oc43/LC506796/HCoV_OC43/Yamagata.JPN/20.2011_1168/2011\n",
      "oc43/KF572823/10290_10/2010\n",
      "oc43/KF572849/5414_07/2007\n",
      "oc43/KF572817/3582_06/2006\n",
      "oc43/KF572818/3647_06/2006\n",
      "oc43/JN129834/HK04_01/2004\n",
      "oc43/KF963235/HCoV_OC43/FRA_EPI/Caen/2004/06/2004\n",
      "oc43/JN129835/HK04_02/2004\n",
      "oc43/KF572832/12691_12/2012\n",
      "oc43/KF572841/3269A_12/2012\n",
      "oc43/KF963240/HCoV_OC43/FRA_EPI/Caen/2009/11/2009\n",
      "oc43/LC506801/HCoV_OC43/Yamagata.JPN/04.2012_0168/2012\n",
      "oc43/LC506802/HCoV_OC43/Yamagata.JPN/05.2012_0279/2012\n",
      "oc43/LC506790/HCoV_OC43/Yamagata.JPN/09.2011_0752/2011\n",
      "oc43/LC506798/HCoV_OC43/Yamagata.JPN/21.2011_1236/2011\n",
      "oc43/LC506808/HCoV_OC43/Yamagata.JPN/13.2012_0719/2012\n",
      "oc43/LC506820/HCoV_OC43/Yamagata.JPN/07.2013_0380/2013\n",
      "oc43/LC506793/HCoV_OC43/Yamagata.JPN/14.2011_0874/2011\n",
      "oc43/LC506812/HCoV_OC43/Yamagata.JPN/51.2012_2581/2012\n",
      "oc43/LC506813/HCoV_OC43/Yamagata.JPN/52.2012_2586/2012\n",
      "oc43/LC506814/HCoV_OC43/Yamagata.JPN/52.2012_2590/2012\n",
      "oc43/LC506791/HCoV_OC43/Yamagata.JPN/13.2011_0825/2011\n",
      "oc43/LC506799/HCoV_OC43/Yamagata.JPN/22.2011_1255/2011\n",
      "oc43/LC506811/HCoV_OC43/Yamagata.JPN/49.2012_2416/2012\n",
      "oc43/LC506816/HCoV_OC43/Yamagata.JPN/05.2013_0300/2013\n",
      "oc43/LC506817/HCoV_OC43/Yamagata.JPN/06.2013_0348/2013\n",
      "oc43/LC506818/HCoV_OC43/Yamagata.JPN/06.2013_0375/2013\n",
      "oc43/KX344031/OC43/human/Mex/LRTI_238/2011/2011\n",
      "oc43/LC506815/HCoV_OC43/Yamagata.JPN/04.2013_0289/2013\n",
      "oc43/LC506795/HCoV_OC43/Yamagata.JPN/16.2011_0969/2011\n",
      "oc43/LC506807/HCoV_OC43/Yamagata.JPN/09.2012_0501/2012\n",
      "oc43/LC506792/HCoV_OC43/Yamagata.JPN/13.2011_0857/2011\n",
      "oc43/LC506800/HCoV_OC43/Yamagata.JPN/27.2011_1525/2011\n",
      "oc43/LC506773/HCoV_OC43/Yamagata.JPN/39.2010_2197/2010\n",
      "oc43/LC506772/HCoV_OC43/Yamagata.JPN/38.2010_2108/2010\n",
      "oc43/LC506803/HCoV_OC43/Yamagata.JPN/05.2012_0283/2012\n",
      "oc43/LC506797/HCoV_OC43/Yamagata.JPN/21.2011_1174/2011\n",
      "oc43/LC506804/HCoV_OC43/Yamagata.JPN/06.2012_0291/2012\n",
      "oc43/LC506806/HCoV_OC43/Yamagata.JPN/06.2012_0371/2012\n",
      "oc43/KF963243/HCoV_OC43/FRA_EPI/Caen/2012/14/2012\n",
      "oc43/KF963244/HCoV_OC43/FRA_EPI/Caen/2013/15/2013\n",
      "oc43/LC506931/HCoV_OC43/Yamagata.JPN/09.2017_0410/2017\n",
      "oc43/KF572831/12689_12/2012\n",
      "oc43/LC331068/HCoV_OC43/Yamagata.JPN/02.15_0022/2015\n",
      "oc43/LC331069/HCoV_OC43/Yamagata.JPN/02.15_0076/2015\n",
      "oc43/LC506845/HCoV_OC43/Yamagata.JPN/02.2015_0077/2015\n",
      "oc43/LC506846/HCoV_OC43/Yamagata.JPN/03.2015_0086/2015\n",
      "oc43/LC506848/HCoV_OC43/Yamagata.JPN/03.2015_0093/2015\n",
      "oc43/LC506849/HCoV_OC43/Yamagata.JPN/03.2015_0101/2015\n",
      "oc43/LC506851/HCoV_OC43/Yamagata.JPN/03.2015_0129/2015\n",
      "oc43/LC506852/HCoV_OC43/Yamagata.JPN/03.2015_0130/2015\n",
      "oc43/LC506853/HCoV_OC43/Yamagata.JPN/03.2015_0131/2015\n",
      "oc43/LC506857/HCoV_OC43/Yamagata.JPN/04.2015_0140/2015\n",
      "oc43/LC331073/HCoV_OC43/Yamagata.JPN/05.15_0234/2015\n",
      "oc43/LC506873/HCoV_OC43/Yamagata.JPN/05.2015_0218/2015\n",
      "oc43/LC506880/HCoV_OC43/Yamagata.JPN/05.2015_0230/2015\n",
      "oc43/LC506882/HCoV_OC43/Yamagata.JPN/05.2015_0233/2015\n",
      "oc43/LC506950/HCoV_OC43/Yamagata.JPN/05.2015_0285/2015\n",
      "oc43/LC506914/HCoV_OC43/Yamagata.JPN/11.2015_0604/2015\n",
      "oc43/LC506895/HCoV_OC43/Yamagata.JPN/06.2015_0300/2015\n",
      "oc43/LC506844/HCoV_OC43/Yamagata.JPN/53.2014_2422/2014\n",
      "oc43/LC506850/HCoV_OC43/Yamagata.JPN/03.2015_0128/2015\n",
      "oc43/LC506854/HCoV_OC43/Yamagata.JPN/03.2015_0134/2015\n",
      "oc43/LC506858/HCoV_OC43/Yamagata.JPN/04.2015_0142/2015\n",
      "oc43/LC506867/HCoV_OC43/Yamagata.JPN/04.2015_0209/2015\n",
      "oc43/LC506870/HCoV_OC43/Yamagata.JPN/04.2015_0213/2015\n",
      "oc43/LC331074/HCoV_OC43/Yamagata.JPN/05.15_0248/2015\n",
      "oc43/LC506871/HCoV_OC43/Yamagata.JPN/05.2015_0215/2015\n",
      "oc43/LC506887/HCoV_OC43/Yamagata.JPN/05.2015_0282/2015\n",
      "oc43/LC506900/HCoV_OC43/Yamagata.JPN/06.2015_0359/2015\n",
      "oc43/LC506904/HCoV_OC43/Yamagata.JPN/07.2015_0373/2015\n",
      "oc43/LC506911/HCoV_OC43/Yamagata.JPN/08.2015_0434/2015\n",
      "oc43/LC506823/HCoV_OC43/Yamagata.JPN/11.2013_0666/2013\n",
      "oc43/LC506916/HCoV_OC43/Yamagata.JPN/19.2015_0902/2015\n",
      "oc43/LC506902/HCoV_OC43/Yamagata.JPN/07.2015_0368/2015\n",
      "oc43/LC506819/HCoV_OC43/Yamagata.JPN/06.2013_0378/2013\n",
      "oc43/LC506822/HCoV_OC43/Yamagata.JPN/10.2013_0627/2013\n",
      "oc43/MF996613/OC43/2015\n",
      "oc43/MF996615/OC43/2015\n",
      "oc43/MF996620/OC43/2014\n",
      "oc43/MF996614/OC43/2015\n",
      "oc43/LC506861/HCoV_OC43/Yamagata.JPN/04.2015_0151/2015\n",
      "oc43/LC506883/HCoV_OC43/Yamagata.JPN/05.2015_0235/2015\n",
      "oc43/LC506886/HCoV_OC43/Yamagata.JPN/05.2015_0279/2015\n",
      "oc43/LC506888/HCoV_OC43/Yamagata.JPN/05.2015_0284/2015\n",
      "oc43/LC506897/HCoV_OC43/Yamagata.JPN/06.2015_0303/2015\n",
      "oc43/LC506912/HCoV_OC43/Yamagata.JPN/08.2015_0486/2015\n",
      "oc43/LC331071/HCoV_OC43/Yamagata.JPN/03.15_0087/2015\n",
      "oc43/LC506847/HCoV_OC43/Yamagata.JPN/03.2015_0088/2015\n",
      "oc43/LC506855/HCoV_OC43/Yamagata.JPN/03.2015_0136/2015\n",
      "oc43/LC331072/HCoV_OC43/Yamagata.JPN/04.15_0214/2015\n",
      "oc43/LC506860/HCoV_OC43/Yamagata.JPN/04.2015_0147/2015\n",
      "oc43/LC506862/HCoV_OC43/Yamagata.JPN/04.2015_0152/2015\n",
      "oc43/LC506863/HCoV_OC43/Yamagata.JPN/04.2015_0154/2015\n",
      "oc43/LC506864/HCoV_OC43/Yamagata.JPN/04.2015_0155/2015\n",
      "oc43/LC506865/HCoV_OC43/Yamagata.JPN/04.2015_0158/2015\n",
      "oc43/LC506869/HCoV_OC43/Yamagata.JPN/04.2015_0211/2015\n",
      "oc43/LC506874/HCoV_OC43/Yamagata.JPN/05.2015_0220/2015\n",
      "oc43/LC506875/HCoV_OC43/Yamagata.JPN/05.2015_0223/2015\n",
      "oc43/LC506876/HCoV_OC43/Yamagata.JPN/05.2015_0224/2015\n",
      "oc43/LC506877/HCoV_OC43/Yamagata.JPN/05.2015_0225/2015\n",
      "oc43/LC506878/HCoV_OC43/Yamagata.JPN/05.2015_0228/2015\n",
      "oc43/LC506879/HCoV_OC43/Yamagata.JPN/05.2015_0229/2015\n",
      "oc43/LC506881/HCoV_OC43/Yamagata.JPN/05.2015_0232/2015\n",
      "oc43/LC506884/HCoV_OC43/Yamagata.JPN/05.2015_0236/2015\n",
      "oc43/LC506890/HCoV_OC43/Yamagata.JPN/05.2015_0288/2015\n",
      "oc43/LC331075/HCoV_OC43/Yamagata.JPN/06.15_0350/2015\n",
      "oc43/LC506891/HCoV_OC43/Yamagata.JPN/06.2015_0290/2015\n",
      "oc43/LC506893/HCoV_OC43/Yamagata.JPN/06.2015_0294/2015\n",
      "oc43/LC506896/HCoV_OC43/Yamagata.JPN/06.2015_0301/2015\n",
      "oc43/LC506901/HCoV_OC43/Yamagata.JPN/07.2015_0361/2015\n",
      "oc43/LC506905/HCoV_OC43/Yamagata.JPN/07.2015_0416/2015\n",
      "oc43/LC506906/HCoV_OC43/Yamagata.JPN/07.2015_0417/2015\n",
      "oc43/LC331078/HCoV_OC43/Yamagata.JPN/11.15_0615/2015\n",
      "oc43/LC506915/HCoV_OC43/Yamagata.JPN/13.2015_0693/2015\n",
      "oc43/LC506841/HCoV_OC43/Yamagata.JPN/46.2014_2103/2014\n",
      "oc43/LC506843/HCoV_OC43/Yamagata.JPN/52.2014_2419/2014\n",
      "oc43/LC506898/HCoV_OC43/Yamagata.JPN/06.2015_0351/2015\n",
      "oc43/LC506913/HCoV_OC43/Yamagata.JPN/09.2015_0492/2015\n",
      "oc43/LC506903/HCoV_OC43/Yamagata.JPN/07.2015_0369/2015\n",
      "oc43/LC506889/HCoV_OC43/Yamagata.JPN/05.2015_0287/2015\n",
      "oc43/LC506907/HCoV_OC43/Yamagata.JPN/07.2015_0418/2015\n",
      "oc43/LC506908/HCoV_OC43/Yamagata.JPN/07.2015_0424/2015\n",
      "oc43/LC506909/HCoV_OC43/Yamagata.JPN/07.2015_0425/2015\n",
      "oc43/LC506910/HCoV_OC43/Yamagata.JPN/08.2015_0433/2015\n",
      "oc43/LC331076/HCoV_OC43/Yamagata.JPN/10.15_0600/2015\n",
      "oc43/LC331077/HCoV_OC43/Yamagata.JPN/11.15_0612/2015\n",
      "oc43/KF572833/12694_12/2012\n",
      "oc43/KY967356/HCoV_OC43/Seattle/USA/SC2924/2015/2015\n",
      "oc43/LC506856/HCoV_OC43/Yamagata.JPN/04.2015_0137/2015\n",
      "oc43/LC506832/HCoV_OC43/Yamagata.JPN/06.2014_0378/2014\n",
      "oc43/LC506834/HCoV_OC43/Yamagata.JPN/08.2014_0432/2014\n",
      "oc43/LC506835/HCoV_OC43/Yamagata.JPN/08.2014_0434/2014\n",
      "oc43/LC506837/HCoV_OC43/Yamagata.JPN/08.2014_0436/2014\n",
      "oc43/LC506859/HCoV_OC43/Yamagata.JPN/04.2015_0144/2015\n",
      "oc43/LC506892/HCoV_OC43/Yamagata.JPN/06.2015_0291/2015\n",
      "oc43/LC506885/HCoV_OC43/Yamagata.JPN/05.2015_0275/2015\n",
      "oc43/LC506836/HCoV_OC43/Yamagata.JPN/08.2014_0435/2014\n",
      "oc43/LC506899/HCoV_OC43/Yamagata.JPN/07.2015_0353/2015\n",
      "oc43/LC506842/HCoV_OC43/Yamagata.JPN/52.2014_2417/2014\n",
      "oc43/MF374983/HCoV_OC43/USA/TCNP_0070/2016/2016\n",
      "oc43/LC331070/HCoV_OC43/Yamagata.JPN/02.15_0083/2015\n",
      "oc43/LC506894/HCoV_OC43/Yamagata.JPN/06.2015_0297/2015\n",
      "oc43/MH121121/HCoV_OC43/USA/ACRI_0213/2016/2016\n",
      "oc43/MF374985/HCoV_OC43/USA/TCNP_00212/2017/2017\n",
      "oc43/KY369907/HCoV_OC43/Seattle/USA/SC9741/2016/2016\n",
      "oc43/MF374984/HCoV_OC43/USA/TCNP_00204/2017/2017\n",
      "oc43/KY983583/HCoV_OC43/Seattle/USA/SC2481/2015/2015\n",
      "oc43/KY983588/HCoV_OC43/Seattle/USA/SC3118/2015/2015\n",
      "oc43/KY369906/HCoV_OC43/Seattle/USA/SC622/2016/2016\n",
      "oc43/KY369905/HCoV_OC43/Seattle/USA/SC831/2016/2016\n",
      "oc43/LC506917/HCoV_OC43/Yamagata.JPN/05.2017_0193/2017\n",
      "oc43/LC506922/HCoV_OC43/Yamagata.JPN/07.2017_0305/2017\n",
      "oc43/LC506926/HCoV_OC43/Yamagata.JPN/07.2017_0361/2017\n",
      "oc43/LC506928/HCoV_OC43/Yamagata.JPN/08.2017_0364/2017\n",
      "oc43/KY983585/HCoV_OC43/Seattle/USA/SC2854/2015/2015\n",
      "oc43/KY967359/HCoV_OC43/Seattle/USA/SC2730/2015/2015\n",
      "oc43/KY967361/HCoV_OC43/Seattle/USA/SC2345/2015/2015\n",
      "oc43/KY684759/HCoV_OC43/Seattle/USA/SC2269/2016/2016\n",
      "oc43/LC506918/HCoV_OC43/Yamagata.JPN/05.2017_0247/2017\n",
      "oc43/LC506919/HCoV_OC43/Yamagata.JPN/06.2017_0266/2017\n",
      "oc43/LC506923/HCoV_OC43/Yamagata.JPN/07.2017_0306/2017\n",
      "oc43/LC506924/HCoV_OC43/Yamagata.JPN/07.2017_0314/2017\n",
      "oc43/LC506925/HCoV_OC43/Yamagata.JPN/07.2017_0354/2017\n",
      "oc43/LC506927/HCoV_OC43/Yamagata.JPN/07.2017_0362/2017\n",
      "oc43/LC506929/HCoV_OC43/Yamagata.JPN/08.2017_0375/2017\n",
      "oc43/LC506930/HCoV_OC43/Yamagata.JPN/08.2017_0379/2017\n",
      "oc43/LC506932/HCoV_OC43/Yamagata.JPN/09.2017_0411/2017\n",
      "oc43/LC506933/HCoV_OC43/Yamagata.JPN/09.2017_0478/2017\n",
      "oc43/LC506934/HCoV_OC43/Yamagata.JPN/09.2017_0480/2017\n",
      "oc43/LC506940/HCoV_OC43/Yamagata.JPN/12.2017_0642/2017\n",
      "oc43/LC506941/HCoV_OC43/Yamagata.JPN/12.2017_0751/2017\n",
      "oc43/LC506943/HCoV_OC43/Yamagata.JPN/16.2017_1199/2017\n",
      "oc43/LC506944/HCoV_OC43/Yamagata.JPN/20.2017_1370/2017\n",
      "oc43/LC506935/HCoV_OC43/Yamagata.JPN/09.2017_0482/2017\n",
      "oc43/LC506951/HCoV_OC43/Yamagata.JPN/10.2017_0491/2017\n",
      "oc43/LC506937/HCoV_OC43/Yamagata.JPN/11.2017_0521/2017\n",
      "oc43/LC506942/HCoV_OC43/Yamagata.JPN/14.2017_0959/2017\n",
      "oc43/LC506921/HCoV_OC43/Yamagata.JPN/07.2017_0304/2017\n",
      "oc43/LC506936/HCoV_OC43/Yamagata.JPN/10.2017_0516/2017\n",
      "oc43/LC331083/HCoV_OC43/Yamagata.JPN/23.16_0980/2016\n",
      "oc43/LC506920/HCoV_OC43/Yamagata.JPN/06.2017_0300/2017\n",
      "oc43/KY967358/HCoV_OC43/Seattle/USA/SC2770/2015/2015\n",
      "oc43/LC506938/HCoV_OC43/Yamagata.JPN/11.2017_0634/2017\n",
      "oc43/LC506945/HCoV_OC43/Yamagata.JPN/26.2017_1566/2017\n",
      "oc43/MF996618/OC43/2015\n",
      "oc43/MN026164/OC43_KLF_01_2018/2018\n",
      "oc43/LC506805/HCoV_OC43/Yamagata.JPN/06.2012_0369/2012\n",
      "oc43/LC506810/HCoV_OC43/Yamagata.JPN/46.2012_2185/2012\n",
      "oc43/KF572816/229_05/2005\n",
      "oc43/MN306036/HCoV_OC43/Seattle/USA/SC0682/2019/2019\n",
      "oc43/MN306041/HCoV_OC43/Seattle/USA/SC0810/2019/2019\n",
      "oc43/MN310478/HCoV_OC43/Seattle/USA/SC0776/2019/2019\n",
      "oc43/MN306042/HCoV_OC43/Seattle/USA/SC0839/2019/2019\n",
      "oc43/MN306053/HCoV_OC43/Seattle/USA/SC9430/2018/2019\n",
      "oc43/MF996608/OC43/2015\n",
      "oc43/MF996616/OC43/2015\n",
      "oc43/MF996617/OC43/2015\n",
      "oc43/KF963231/HCoV_OC43/FRA_EPI/Caen/2001/02/2001\n",
      "oc43/KF530068/OC43/human/USA/007_11/2000/2000\n",
      "oc43/KF530070/OC43/human/USA/991_19/1999/1999\n",
      "oc43/KF530081/OC43/human/USA/991_5/1999/1999\n",
      "oc43/KF530088/OC43/human/USA/901_54/1990/1990\n",
      "oc43/KF530091/OC43/human/USA/911_58/1991/1991\n",
      "oc43/KF530071/OC43/human/USA/925_1/1992/1992\n",
      "oc43/KF530076/OC43/human/USA/911_11/1991/1991\n",
      "oc43/KF530096/OC43/human/USA/911_38/1991/1991\n",
      "oc43/KF530095/OC43/human/USA/912_6/1991/1991\n",
      "oc43/KF530089/OC43/human/USA/911_66/1991/1991\n",
      "oc43/KF530082/OC43/human/USA/912_11/1991/1991\n",
      "oc43/KF530094/OC43/human/USA/912_36/1991/1991\n",
      "oc43/KF530079/OC43/human/USA/913_29/1991/1991\n",
      "oc43/KF530067/OC43/human/USA/912_10/1991/1991\n",
      "oc43/KF530084/OC43/human/USA/951_18/1995/1995\n",
      "oc43/KF530063/OC43/human/USA/9612_48/1996/1996\n",
      "oc43/KF530069/OC43/human/USA/982_4/1998/1998\n",
      "oc43/KF530099/OC43/human/USA/971_5/1997/1997\n",
      "oc43/KF530098/OC43/human/USA/965_6/1996/1996\n",
      "oc43/KF963232/HCoV_OC43/FRA_EPI/Caen/2002/03/2002\n",
      "oc43/KF963229/HCoV_OC43/FRA_EPI/Caen/1967/VR759/1967\n"
     ]
    }
   ],
   "source": [
    "syn_nonsyn_histogram('oc43', 'spike', 'A')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def divergence_by_clade(cov, gene, window, clade, min_seqs, year_max=None, year_min=None):\n",
    "    #Find fraction of sites that differ from root and average among all viruses at each time point\n",
    "    input_file_root = '../'+str(cov)+'/auspice/seasonal_corona_'+str(cov)+'_'+str(gene)+'_root-sequence.json'\n",
    "    input_file_alignment = '../'+str(cov)+'/results/aligned_'+str(cov)+'_'+str(gene)+'.fasta'\n",
    "    metafile = '../'+str(cov)+'/results/metadata_'+str(cov)+'_'+str(gene)+'.tsv'\n",
    "    \n",
    "    #Subset data based on time windows\n",
    "    meta = pd.read_csv(metafile, sep = '\\t')\n",
    "    meta.drop(meta[meta['date']=='?'].index, inplace=True)\n",
    "    meta.dropna(subset=['date'], inplace=True)\n",
    "    meta['year'] = meta['date'].str[:4].astype('int')\n",
    "    \n",
    "    if year_max:\n",
    "        meta.drop(meta[meta['year']>year_max].index, inplace=True)\n",
    "    if year_min:\n",
    "        meta.drop(meta[meta['year']<year_min].index, inplace=True)\n",
    "    \n",
    "    date_range = meta['year'].max() - meta['year'].min()\n",
    "    \n",
    "    \n",
    "    if clade!= None:\n",
    "        clade_df = separate_clades(cov, gene)\n",
    "        meta = meta.merge(clade_df, on='strain')\n",
    "        meta.drop(meta[meta['clade']!=clade].index, inplace=True)\n",
    "\n",
    "    #Group viruses by time windows\n",
    "    virus_time_subset = {}\n",
    "    if window == 'all':\n",
    "        years = str(meta['year'].min()) + '-' + str(meta['year'].max())\n",
    "        virus_time_subset[years] = meta['strain'].tolist()\n",
    "    else:\n",
    "        date_window_start = meta['year'].min()\n",
    "        date_window_end = meta['year'].min() + window\n",
    "        while date_window_end <= meta['year'].max():\n",
    "            years = str(date_window_start) + '-' + str(date_window_end)\n",
    "            strains = meta[(meta['year']>=date_window_start) & (meta['year']<date_window_end)]['strain'].tolist()\n",
    "            virus_time_subset[years] = strains\n",
    "            #sliding window\n",
    "            date_window_end += 1\n",
    "            date_window_start += 1     \n",
    "\n",
    "    #initiate lists to record all time windows\n",
    "    year_windows = []\n",
    "    seqs_in_window = []\n",
    "    total_subs_fractions = []\n",
    "    syn_fractions = []\n",
    "    nonsyn_fractions = []\n",
    "    total_subs_fractions_avg = []\n",
    "    syn_fractions_avg = []\n",
    "    nonsyn_fractions_avg = []\n",
    "    \n",
    "    #track all mutations that occur in tree\n",
    "    nt_muts = []\n",
    "    \n",
    "    #strains at first time point(to make consensus from)\n",
    "    first_window = True\n",
    "    first_window_strains = []\n",
    "    first_window_sequences = []\n",
    "    root_seq = ''\n",
    "    root_aa_seq = ''\n",
    "    \n",
    "    for years, subset_viruses in virus_time_subset.items():\n",
    "        #don't use windows with fewer than min_seqs \n",
    "        if len(subset_viruses) >= min_seqs:\n",
    "            year_windows.append(years)\n",
    "            seqs_in_window.append(len(subset_viruses))\n",
    "    \n",
    "            \n",
    "            #make consensus sequence at first time point\n",
    "            if first_window == True:\n",
    "                first_window_strains+=subset_viruses\n",
    "                with open(input_file_alignment, \"r\") as aligned_handle:\n",
    "                    for virus in SeqIO.parse(aligned_handle, \"fasta\"):\n",
    "                        if virus.id in first_window_strains:\n",
    "                            first_window_sequences.append(virus)\n",
    "                            \n",
    "                first_window_alignment = MultipleSeqAlignment(first_window_sequences)\n",
    "                root_seq = AlignInfo.SummaryInfo(first_window_alignment).dumb_consensus(ambiguous ='N')\n",
    "                root_aa_seq = root_seq.translate()\n",
    "                \n",
    "                first_window = False\n",
    "            \n",
    "        \n",
    "        \n",
    "            \n",
    "            fraction_subs = []\n",
    "            fraction_syn = []\n",
    "            fraction_nonsyn = []\n",
    "            \n",
    "            #calculate divergence at all subsequent time points\n",
    "            with open(input_file_alignment, \"r\") as aligned_handle:\n",
    "                for virus in SeqIO.parse(aligned_handle, \"fasta\"):         \n",
    "#                     Only viruses in time window\n",
    "                    if virus.id in subset_viruses:    \n",
    "                        \n",
    "                        #check\n",
    "                        if len(virus.seq) != len(root_seq):\n",
    "                            print(virus)\n",
    "                        elif len(virus.seq) == len(root_seq):\n",
    "                            count_total_unambiguous = 0\n",
    "                            count_subs = 0\n",
    "                            count_syn_subs = 0\n",
    "                            count_nonsyn_subs = 0\n",
    "                            for pos in range(len(root_seq)):\n",
    "                                root_nt = str(root_seq[pos])\n",
    "                                virus_nt = str(virus.seq[pos])\n",
    "                                #skip ambiguous sites\n",
    "                                if virus_nt != 'N':\n",
    "                                    if root_nt != 'N':\n",
    "                                        count_total_unambiguous+=1\n",
    "                                        if virus_nt != root_nt:\n",
    "                                            count_subs+=1\n",
    "                                            #determine syn or nonsyn\n",
    "                                            codon = math.floor(pos/3)\n",
    "                                            codon_pos = pos-(codon*3)\n",
    "                                            if codon_pos == 0:\n",
    "                                                codon_nt = virus.seq[pos:(pos+3)]\n",
    "                                            elif codon_pos == 1:\n",
    "                                                codon_nt = virus.seq[(pos-1):(pos+2)]\n",
    "                                            elif codon_pos == 2:\n",
    "                                                codon_nt = virus.seq[(pos-2):(pos+1)]\n",
    "                                            codon_aa = codon_nt.translate()\n",
    "                                            root_aa = root_aa_seq[codon]\n",
    "                                            #skip ambiguous\n",
    "                                            if root_aa != 'X':\n",
    "                                                if codon_aa != root_aa:\n",
    "                                                    count_nonsyn_subs+=1\n",
    "                                                    nt_mut = str(root_nt)+str(pos)+str(virus_nt)+'nonsyn'\n",
    "                                                elif codon_aa == root_aa:\n",
    "                                                    count_syn_subs+=1\n",
    "                                                    nt_mut = str(root_nt)+str(pos)+str(virus_nt)+'syn'\n",
    "                                                if nt_mut not in nt_muts:\n",
    "                                                    nt_muts.append(nt_mut)\n",
    "                            fraction_sub_virus = count_subs/count_total_unambiguous\n",
    "                            fraction_subs.append(fraction_sub_virus)\n",
    "                            fraction_syn_virus = count_syn_subs/count_total_unambiguous\n",
    "                            fraction_syn.append(fraction_syn_virus)\n",
    "                            fraction_nonsyn_virus = count_nonsyn_subs/count_total_unambiguous\n",
    "                            fraction_nonsyn.append(fraction_nonsyn_virus)\n",
    "\n",
    "\n",
    "\n",
    "            fraction_subs_window = sum(fraction_subs)/len(fraction_subs)\n",
    "            fraction_syn_window = sum(fraction_syn)/len(fraction_syn)\n",
    "            fraction_nonsyn_window = sum(fraction_nonsyn)/len(fraction_nonsyn)\n",
    "            \n",
    "            total_subs_fractions.append(fraction_subs)\n",
    "            syn_fractions.append(fraction_syn)\n",
    "            nonsyn_fractions.append(fraction_nonsyn)\n",
    "            \n",
    "            total_subs_fractions_avg.append(fraction_subs_window)\n",
    "            syn_fractions_avg.append(fraction_syn_window)\n",
    "            nonsyn_fractions_avg.append(fraction_nonsyn_window)\n",
    "            \n",
    "    return year_windows, seqs_in_window, total_subs_fractions, syn_fractions, nonsyn_fractions, total_subs_fractions_avg, syn_fractions_avg, nonsyn_fractions_avg, nt_muts\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
